test/src/test/java/hudson/model/QueueTest.java

```
@@ -780,4 +786,108 @@ public void run() {
+        matrixProject.setAssignedLabel(label);
+        FreeStyleProject upstreamProject = r.createFreeStyleProject("upstream");
+        //let's assume the flyweighttask has an upstream project and that must be blocked
+        // when the upstream project is running
+        matrixProject.addTrigger(new ReverseBuildTrigger("upstream", Result.SUCCESS));
+        matrixProject.setBlockBuildWhenUpstreamBuilding(true);
+        //we schedule the project but we pretend no executors are available thus
+        //the flyweight task is in the buildable queue without being executed
+        matrixProject.scheduleBuild2(0);
+        //let s wait for the Queue instance to be updated
+        while (Queue.getInstance().getBuildableItems().size() != 1) {
+            Thread.sleep(10);
+        }
+        //in this state the build is not blocked, it's just waiting for an available executor
+        assertFalse(Queue.getInstance().getItems()[0].isBlocked());
+        //we start the upstream project that should block the downstream one
+        upstreamProject.getBuildersList().add(new SleepBuilder(10000));
+        upstreamProject.setDisplayName("upstream");
+        QueueTaskFuture upstream = upstreamProject.scheduleBuild2(0);
+        //let s wait for the Upstream to enter the buildable Queue
+        boolean enteredTheQueue = false;
+        while (!enteredTheQueue) {
+            for (Queue.BuildableItem item : Queue.getInstance().getBuildableItems()) {
+                if (item.task.getDisplayName() != null && item.task.getDisplayName().equals(upstreamProject.getDisplayName())) {
+                    enteredTheQueue = true;
+                }
+            }
+        }
+        //let's wait for the upstream project to actually start so that we're sure the Queue has been updated
+        //when the upstream starts the downstream has already left the buildable queue and the queue is empty
+        while (!Queue.getInstance().getBuildableItems().isEmpty()) {
```

should I add a timeout and let the test fail if the builds gets somehow blocked?